services/simcore-agent/tests/unit/test_volumes_cleanup_s3.py

```diff
@@ -0,0 +1,170 @@
+# pylint: disable=redefined-outer-name
+# pylint: disable=unused-argument
+# pylint: disable=too-many-locals
+
+import hashlib
+from pathlib import Path
+from typing import Optional
+
+import aioboto3
+import pytest
+from aiodocker.volumes import DockerVolume
+from pydantic import HttpUrl
+from simcore_service_simcore_agent._settings import ApplicationSettings
+from simcore_service_simcore_agent.volumes_cleanup._s3 import (
+    S3Provider,
+    _get_s3_path,
+    store_to_s3,
+)
+
+# UTILS
+
+
+def _get_file_hashes_in_path(
+    path_to_hash: Path, exclude_files: Optional[set[Path]] = None
+) -> set[tuple[Path, str]]:
+    def _hash_path(path: Path):
+        sha256_hash = hashlib.sha256()
+        with open(path, "rb") as file:  # pylint:disable=unspecified-encoding
+            # Read and update hash string value in blocks of 4K
+            for byte_block in iter(lambda: file.read(4096), b""):
+                sha256_hash.update(byte_block)
+        return sha256_hash.hexdigest()
+
+    def _relative_path(root_path: Path, full_path: Path) -> Path:
+        return full_path.relative_to(root_path)
+
+    if path_to_hash.is_file():
+        return {(_relative_path(path_to_hash, path_to_hash), _hash_path(path_to_hash))}
+
+    if exclude_files is None:
+        exclude_files = set()
+
+    return {
+        (_relative_path(path_to_hash, path), _hash_path(path))
+        for path in path_to_hash.rglob("*")
+        if path.is_file() and _relative_path(path_to_hash, path) not in exclude_files
+    }
+
+
+async def _download_files_from_bucket(
+    endpoint: str,
+    access_key: str,
+    secret_key: str,
+    bucket_name: str,
+    save_to: Path,
+    swarm_stack_name: str,
+    study_id: str,
+    node_uuid: str,
+    run_id: str,
+) -> None:
+    session = aioboto3.Session(
+        aws_access_key_id=access_key, aws_secret_access_key=secret_key
+    )
+    async with session.resource("s3", endpoint_url=endpoint, use_ssl=False) as s_3:
+        bucket = await s_3.Bucket(bucket_name)
+        async for s3_object in bucket.objects.all():
+            key_path = f"{swarm_stack_name}/{study_id}/{node_uuid}/{run_id}/"
+            if s3_object.key.startswith(key_path):
+                file_object = await s3_object.get()
+                file_path: Path = save_to / s3_object.key.replace(key_path, "")
+                file_path.parent.mkdir(parents=True, exist_ok=True)
+                print(f"Saving file to {file_path}")
+                file_content = await file_object["Body"].read()
+                file_path.write_bytes(file_content)
+
+
+def _create_data(folder: Path) -> None:
```

Comment:
are you aware of file_extra.py in pytest-simcore? please check there. I think you could re-use it here